external/fv3fit/fv3fit/keras/_models/recurrent.py

```diff
@@ -0,0 +1,616 @@
+            n_units: number of neurons in each hidden layer
+            n_hidden_layers: number of hidden layers
+            kernel_regularizer: if given, use for regularizing
+                hidden layer kernels
+            train_batch_size: batch size to use for keras model training
+            optimizer: optimizer for keras training
+            use_moisture_limiter: if True, prevent the moisture tendency from
+                depleting all moisture on its own in one training timestep.
+                Moisture can still go negative if there is a negative
+                physics tendency of moisture.
+            state_noise: amount of Gaussian noise to add to model prognostic state
+                before using it to predict tendencies
+        """
+        prognostic_variables = (
+            "air_temperature",
+            "specific_humidity",
+        )
+        self.given_tendency_names = (
+            "air_temperature_tendency_due_to_model",
+            "specific_humidity_tendency_due_to_model",
+        )
+        output_variables = ["dQ1", "dQ2"]
+        super(BPTTModel, self).__init__(
+            sample_dim_name, input_variables, output_variables,
+        )
+        self.input_packer = ArrayPacker(sample_dim_name, input_variables)
+        self.prognostic_packer = ArrayPacker(sample_dim_name, prognostic_variables)
+        self.train_timestep_seconds: Optional[float] = None
+        self.input_scaler = LayerStandardScaler()
+        self.prognostic_scaler = LayerStandardScaler()
+        self.tendency_scaler = LayerStandardScaler()
+        self.n_units = n_units
+        self.n_hidden_layers = n_hidden_layers
+        # self.tendency_ratio = tendency_ratio
+        self.kernel_regularizer = kernel_regularizer
+        self.train_batch_size = train_batch_size
+        self.optimizer = optimizer
+        self.activation = "relu"
+        self.train_keras_model: Optional[tf.keras.Model] = None
+        self.train_tendency_model: Optional[tf.keras.Model] = None
+        self.predict_keras_model: Optional[tf.keras.Model] = None
+        self.use_moisture_limiter = use_moisture_limiter
+        self.state_noise = state_noise
+        self._stepwise_output_metadata: List[DataArrayMetadata] = []
+
+    def build_for(self, X: xr.Dataset):
+        """
+        Given a dataset with [sample, time, z] and [sample, time] arrays
+        containing data in sample windows, fit the scalers and packers
+        and define the keras model (e.g. with the given window length).
+        Timestep is determined from the difference between the first two
+        times in the dataset.
+        """
+        if self.train_keras_model is not None:
+            raise RuntimeError("cannot build, model is already built!")
+        inputs = self.input_packer.to_array(X, is_3d=True)
+        state = self.prognostic_packer.to_array(X, is_3d=True)
+        self.input_scaler.fit(inputs)
+        self.prognostic_scaler.fit(state)
```

Comment:
Can  you revert 497d6ed3973fa95611a39681245eecdc062a5203 if you use this suggestion:
```suggestion
        inputs = self.input_packer.to_array(X)
        state = self.prognostic_packer.to_array(X)
        self.input_scaler.fit(inputs)
        self.prognostic_scaler.fit(state)
```

Reply:
I think I can if I also collapse `time` and `sample` into one dimension before calling this fit routine within the `BPTTModel`. I'll look into it.